nifi-nar-bundles/nifi-windows-event-log-bundle/nifi-windows-event-log-processors/src/main/java/org/apache/nifi/processors/windows/event/log/EvtSubscribe.java

```diff
@@ -0,0 +1,272 @@
+        int maxEventQueueSize = context.getProperty(MAX_EVENT_QUEUE_SIZE).asInteger();
+        evtSubscribeCallback = new EventSubscribeXmlRenderingCallback(getLogger(), s -> {
+            ProcessSessionFactory processSessionFactory = sessionFactoryReference.get();
+            if (processSessionFactory == null) {
+                addRenderedXml(s, maxEventQueueSize);
+            } else {
+                ProcessSession session = processSessionFactory.createSession();
+                createAndTransferEventFlowFile(session, s);
+                session.commit();
+            }
+        }, context.getProperty(MAX_BUFFER_SIZE).asInteger(), wEvtApi, kernel32);
+        subscriptionHandle = wEvtApi.EvtSubscribe(null, null,
+                context.getProperty(CHANNEL).getValue(), context.getProperty(QUERY).getValue(), null, null,
+                evtSubscribeCallback, WEvtApi.EvtSubscribeFlags.SUBSCRIBE_TO_FUTURE);
+    }
+
+    /**
+     * Adds a rendered xml to the queue to be processed.
+     * Should only be called within block synchronized on evtSubscribeCallback before the sessionFactoryReference is set
+     *
+     * @param s the rendered xml string
+     * @param maxEventQueueSize the maximum size of the event queue
+     */
+    protected void addRenderedXml(String s, int maxEventQueueSize) {
+        renderedXMLs.add(s);
+        while (renderedXMLs.size() > maxEventQueueSize) {
+            renderedXMLs.poll();
+        }
+    }
+
+    /**
+     * Cleanup native subscription
+     */
+    @OnStopped
+    public void closeSubscriptionHandle() {
+        kernel32.CloseHandle(subscriptionHandle);
+        subscriptionHandle = null;
+    }
+
+    /**
+     * Creates a flow file from the given XML and transfers it to the success relationship
+     *
+     * @param session the process session
+     * @param xml the XML
+     */
+    protected void createAndTransferEventFlowFile(ProcessSession session, String xml) {
+        FlowFile flowFile = session.create();
+        flowFile = session.write(flowFile, out -> out.write(xml.getBytes(Charsets.UTF_8)));
+        flowFile = session.putAttribute(flowFile, CoreAttributes.MIME_TYPE.key(), MediaType.APPLICATION_XML_UTF_8.toString());
+        session.transfer(flowFile, REL_SUCCESS);
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFactory) throws ProcessException {
+        ProcessSession session = sessionFactory.createSession();
+
+        // Give callback the information it needs to do its job and empty queue of events from before trigger
+        synchronized (evtSubscribeCallback) {
+            sessionFactoryReference.compareAndSet(null, sessionFactory);
+            String renderedXml;
+            while ((renderedXml = renderedXMLs.poll()) != null) {
+                createAndTransferEventFlowFile(session, renderedXml);
+            }
+            session.commit();
+        }
+        context.yield();
```

Comment:
Should definitely not yield after every onTrigger. That defeats the purpose of yielding and the scheduling period